types/datum.go

```diff
@@ -1531,7 +1534,11 @@ func (d *Datum) toSignedInteger(sc *stmtctx.StatementContext, tp byte) (int64, e
 		return ConvertJSONToInt(sc, d.GetMysqlJSON(), false)
 	case KindBinaryLiteral, KindMysqlBit:
 		val, err := d.GetBinaryLiteral().ToInt(sc)
-		return int64(val), errors.Trace(err)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		ival, err := ConvertUintToInt(val, upperBound, tp)
```

Comment:
Is this behavior the same as MySQL? I think the original logic is more reasonable in this case.

Reply:
This logic is to deal with inserting a value into table, we need to check upperBound.